cmake/modules/AddCephTest.cmake

```diff
@@ -74,7 +74,7 @@ function(add_tox_test name)
   string(REPLACE ";" "," tox_envs "${tox_envs}")
   add_custom_command(
     OUTPUT ${venv_path}/bin/activate
-    COMMAND ${CMAKE_SOURCE_DIR}/src/tools/setup-virtualenv.sh ${venv_path}
+    COMMAND ${CMAKE_SOURCE_DIR}/src/tools/setup-virtualenv.sh --python=python${Python_VERSION_MAJOR}.${Python_VERSION_MINOR} ${venv_path}
```

Comment:
i agree that we cannot assume the existence of python2.7 on the machine where `make check` runs.

but can you help me understand why we are using `python${Python_VERSION_MAJOR}.${Python_VERSION_MINOR}` here?

as `Python_VERSION_MAJOR` and `Python_VERSION_MINOR` are set by the latest `find_package(Python ...)` call. and `add_tox_test()` could be called anywhere in the source tree on demand. 

can we reply on the fact that `find_package(Python ...)` is called before `add_tox_test()` ? and the venv used by tox are used solely for bootstrap tox env, so any python version supporting tox would work. so i think we might want to just put
```cmake
find_package(Python REQUIRED)
add_custom_command(
    OUTPUT ${venv_path}/bin/activate
    COMMAND ${CMAKE_SOURCE_DIR}/src/tools/setup-virtualenv.sh --python="${Python_EXECUTABLE}" "${venv_path}"
    WORKING_DIRECTORY ${tox_path}
    COMMENT "preparing venv for ${name}")
```

also, the commit message does not make sense. please note, the help messages of these options read:
- `WITH_PYTHON2` "build python2 bindings"
- `WITH_PYTHON3` "build python2 bindings"

i don't think any of them is connected to the python version we used for preparing the venv for tox for testing. not to mention `MGR_PYTHON_VERSION`.

Reply:
@tchaikov
I saw that `find_package(Python ...)` is called rather late in the whole configuration. It has given me some doubts already on previous instances.
Perhaps more sensible to move this forward, since with python nothing works in the end.
And put it in `./CMakeLists.txt`

For installing the tox-env with 2.7 when actually the build is for 3.6....
When that happens, I think I saw 2.7 modules being pulled in.

Bottom line is that this patch just makes sure that things are done as asked by the user.

So how about moving the `find_package(Python ...)` stuff up the tree?

Reply:
> Perhaps more sensible to move this forward, since with python nothing works in the end.
And put it in `./CMakeLists.txt`

why?

> done as asked by the user.

please define what is asked by the user.

> So how about moving the `find_package(Python ...)` stuff up the tree?

i'd prefer `find_package()` where it's used.

Reply:
@tchaikov
I would move it up, to make sure that whenever there is a reference to a Python variable that Python detection has been done.  Just look in `./CMakeLists.txt` and see there are lots of packages detected (too) early. 
And like I said, you cannot build ceph without python
So I'm suggesting to move it up, to prevent you earlierer worries about referencing `Python_VERSION_*` without it being detected.

As for this actual PR, 
The user wants his stuff to build with the indicated python version.
In that sense it is silly/confusing to use something different for building some tools.
So perhaps build the env with 2.7 works, when 3.6 was asked.
But it is certainly confusing, and when there are errors, the kind of "oddities" only distract from getting to the real problem. So why not fix it?

Reply:
> to make sure that whenever there is a reference to a Python variable that Python detection has been done. 

i think you need to be more specific. 

> Just look in `./CMakeLists.txt` and see there are lots of packages detected (too) early. 

i don't think we have to follow that it is. otherwise why are we improving Ceph?

> So I'm suggesting to move it up, to prevent you earlierer worries about referencing `Python_VERSION_*` without it being detected.

again, i don't want to search upwards/backwards to make sure that a variable is defined.

> The user wants his stuff to build with the indicated python version. In that sense it is silly/confusing to use something different for building some tools. So perhaps build the env with 2.7 works, when 3.6 was asked.

i think i've explain in https://github.com/ceph/ceph/pull/29761#discussion_r317050077. but probably you missed it?

> it is certainly confusing

please be specific what the confusion is.

Reply:
> > The user wants his stuff to build with the indicated python version. In that sense it is silly/confusing to use something different for building some tools. So perhaps build the env with 2.7 works, when 3.6 was asked.
> 
> i think i've explain in [#29761 (comment)](https://github.com/ceph/ceph/pull/29761#discussion_r317050077). but probably you missed it?

Ah, you want to use  `Python_EXECUTABLE`, that fell out of the scope of the window.
So it looked the same, but was not. That I can fix.

> > it is certainly confusing
> 
> please be specific what the confusion is.

I do not expect any references to python2.7 when I have said `WITH_PYTHON2=OFF` like:
```
Running virtualenv with interpreter /usr/local/bin/python2.7
New python executable in /home/jenkins/workspace/ceph-master36/build/mgr-ansible-virtualenv/bin/python2.7
Also creating executable in /home/jenkins/workspace/ceph-master36/build/mgr-ansible-virtualenv/bin/python
```
Or how about this one:
```
Requirement already satisfied: pathlib2>=2.2.0; python_version < "3.6" in /home/jenkins/workspace/ceph-master36/build/mgr-dashboard-virtualenv/lib/python2.7/site-packages (from pytest<4->-r requirements-test.txt (line 2)) (2.3.4)
```
the part with `python_version < "3.6"`  really makes me wonder.

Reply:
> > to make sure that whenever there is a reference to a Python variable that Python detection has been done.
> 
> i think you need to be more specific.
> 
> > Just look in `./CMakeLists.txt` and see there are lots of packages detected (too) early.
> 
> i don't think we have to follow that it is. otherwise why are we improving Ceph?
> 
> > So I'm suggesting to move it up, to prevent you earlierer worries about referencing `Python_VERSION_*` without it being detected.
> 
> again, i don't want to search upwards/backwards to make sure that a variable is defined.

What you suggested was putting a `find_package(Python....) ` in `cmake/modules/AddCephTest.cmake `??
That would obfuscate to point where the check for python is really made.

I do not really expect those kinds of checks/finds to be hidden in a module.
But if that is what you prefer, we can move (or add) it to `cmake/modules/AddCephTest.cmake `

Reply:
> > The user wants his stuff to build with the indicated python version. In that sense it is silly/confusing to use something different for building some tools. So perhaps build the env with 2.7 works, when 3.6 was asked.
> 
> i think i've explain in [#29761 (comment)](https://github.com/ceph/ceph/pull/29761#discussion_r317050077). but probably you missed it?

Ah, you want to use  `Python_EXECUTABLE`, that fell out of the scope of the window.
So it looked the same, but was not. That I can fix

> > it is certainly confusing
> 
> please be specific what the confusion is.

Reply:
@tchaikov
Is this now to your liking?